# Villes du monde - Données et cartographie #

**Objectif :** afficher une carte avec les 20 plus grandes villes :  

<div class="alert alert-info" role="alert">
  <strong>Travail à faire sur le notebook : </strong> <br>
    Pour les cellules suivantes, appuyez sur <strong>SHIFT + ENTREE </strong>pour lancer le code Python contenu chaque cellule
</div>

# 1 - Importer les données

Le premier travail est d'importer les données sur les villes.  
Les données sont stockées dans un fichier de type tableur (extension .csv).

On commence par importer les bibliothèques utiles :

In [1]:
import csv 

In [2]:
with open('worldcities-top100.csv',newline='') as file:         #Ouverture du fichier csv
    data=[]
    lire=csv.reader(file)                            #chargement des lignes du fichier csv
    print('',end='\n')
    print('Affichage des lignes du tableau',end='\n')
    for ligne in lire:                            #Pour chaque ligne... 
        print(ligne, end='\n')                    #...affichage de la ligne
        data.append(ligne)                     #...on ajoute la ligne dans la liste ...de listes nommée data


Affichage des lignes du tableau
['city', 'lat', 'lng', 'country', 'population', 'id']
['Tokyo', '35.6897', '139.6922', 'Japan', '37977000', '1392685764']
['Jakarta', '-6.2146', '106.8451', 'Indonesia', '34540000', '1360771077']
['Delhi', '28.6600', '77.2300', 'India', '29617000', '1356872604']
['Mumbai', '18.9667', '72.8333', 'India', '23355000', '1356226629']
['Manila', '14.6000', '120.9833', 'Philippines', '23088000', '1608618140']
['Shanghai', '31.1667', '121.4667', 'China', '22120000', '1156073548']
['São Paulo', '-23.5504', '-46.6339', 'Brazil', '22046000', '1076532519']
['Seoul', '37.5600', '126.9900', 'Korea, South', '21794000', '1410836482']
['Mexico City', '19.4333', '-99.1333', 'Mexico', '20996000', '1484247881']
['Guangzhou', '23.1288', '113.2590', 'China', '20902000', '1156237133']
['Beijing', '39.9050', '116.3914', 'China', '19433000', '1156228865']
['Cairo', '30.0561', '31.2394', 'Egypt', '19372000', '1818253931']
['New York', '40.6943', '-73.9249', 'United States', '187

**Questions :** 

1) Quels sont les descipteurs ?  
2) Quelle est la population de Paris ?  
2) Quelle est la ville la plus peuplée ?  
3) Quelle est la première ville des USA ?  



<div class="alert alert-warning" role="alert">
    <strong> Travail à faire sur la fiche réponse : </strong> <br>
    --> Répondre aux questions<br>
</div>

# 2 - Comprendre la structure des données

Les données sont stockées dans une liste de listes.
le premier élément d'une liste a l'indice zéro (en informatique, on commence à compter à zéro). 

In [4]:
data[1]

['Tokyo', '35.6897', '139.6922', 'Japan', '37977000', '1392685764']

In [5]:
data[1][0] # pour accéder au premier élément de la sous liste n°2 

'Tokyo'

**Questions :** 

1) Que fait l'instruction `data[1]` ?  
2) Que contient la première ligne du tableau de données (indice 0) ?  
3) Quelle information apporte la première entrée d'une sous liste (indice 0) ?  
4) Comment obtenir la population de la ville étudiée ?
5) Quel code écrire pour obtenir la population de la 10e ville la plus peuplée ? ( Guangzhou avec 20,9 millions d'habitants ?)  
6) Quel code écrire pour obtenir la latitude de la 5e ville la plus peuplée ? (Manila : lat = 14.6000)  
7) Quel code écrire pour obtenir la longitude de la 5e ville la plus peuplée ? (Manila : lon = 120.9833)  



In [6]:
# tester votre code ici :
data[1][0]

'Tokyo'

<div class="alert alert-warning" role="alert">
    <strong> Travail à faire sur la fiche réponse : </strong> <br>
    --> Répondre aux questions<br>
   
</div>

# 3 - Affichage des cartes sur la ville

On veut maintenant afficher les 20 premières villes de la liste sur une carte :

In [7]:
import folium

In [8]:
carte= folium.Map(location=[0, 0],tiles='Stamen Terrain', zoom_start=2)

<div class="alert alert-warning" role="alert">
    <strong> Travail à faire sur la fiche réponse : </strong> <br>
    --> Lire le code ci-dessous<br>
    --> Corriger le code pour qu'il place les 20 premières villes sur la carte<br>
    --> Recopier le code corrigé sur votre feuille réponse<br>
   
</div>

In [9]:
for i in range(1,10):
    ville = data[i][0]
    population = data[i][ ??? ]
    text = ville + '  ' + str(population)
    
    latitude = data[i][ ??? ]
    longitude = data[i][ ??? ]
    print( text)
    folium.Marker( [latitude,  longitude ], popup=text).add_to(carte)
carte.save('maCarte.html') # pour sauvegarder la carte
carte # pour afficher la carte 

SyntaxError: invalid syntax (<ipython-input-9-3e5eb1d6afe6>, line 3)

In [10]:
# correction

for i in range(1,21):
    ville = data[i][0]
    population = data[i][ 4 ]
    text = ville + '  ' + str(population)
    
    latitude = data[i][ 1 ]
    longitude = data[i][ 2 ]
    print( text)
    folium.Marker( [latitude,  longitude ], popup=text).add_to(carte)
carte.save('maCarte.html') # pour sauvegarder la carte
carte # pour afficher la carte 

Tokyo  37977000
Jakarta  34540000
Delhi  29617000
Mumbai  23355000
Manila  23088000
Shanghai  22120000
São Paulo  22046000
Seoul  21794000
Mexico City  20996000
Guangzhou  20902000
Beijing  19433000
Cairo  19372000
New York  18713220
Kolkāta  17560000
Moscow  17125000
Bangkok  17066000
Buenos Aires  16157000
Shenzhen  15929000
Dhaka  15443000
Lagos  15279000


# 4 - Améliorations

On peut améliorer avec des cercles qui correspondent à la population

On peut changer l'icône du marqueur en ajoutant un paramètre `icon` ainsi:

`icon=folium.Icon(icon='home', color='blue')` 

Pour changer l'icône, remplacer `home`
 des noms suivants :
`cloud` `music` `heart` `home` `user` `film` `cog` `road` `time` `ok` `remove` `play` `download` `picture` `question-sign` `fire` `plane` `thumbs-upwrenc` `glob` `eur` `us` 
`signal` `tree-conifer` `phone` `envelope`

La syntaxe pour afficher un cercle est la suivante :  
`folium.CircleMarker(location=[latitude, longitude], radius=rayon, popup=text, fill=True, fill_color='Red').add_to(carte2)`


**Votre travail :**  
--> changer l'icone des villes  
--> dessiner un cercle dont le rayon représentant la population de la ville

<div class="alert alert-warning" role="alert">
    <strong> Travail à faire sur la fiche réponse : </strong> <br>
    --> Lire le code ci-dessous<br>
    --> Corriger le code pour qu'il place les 20 premières villes sur la carte<br>
    --> Changer l'icone par une étoile<br>      
    --> Tracer un cercle bleu avec un rayon bien choisi ( ni trop grand ni trop petit) mais proportionnel à la population de la ville !<br>  
    --> Recopier le code corrigé sur votre feuille réponse<br>
   
</div>

In [ ]:
# code à lire et à compléter :
carte2= folium.Map(location=[0, 0],tiles='Stamen Terrain', zoom_start=2)

for i in range(1,11):
    ville = data[i][0]
    population = int (data[i][4]) # int pour avoir un nombre
    text = ville + '  ' + str(population) # str pour avoir du texte à écrire
    
    latitude = data[i][ 2 ]
    longitude = data[i][ 1 ]
    folium.Marker( [latitude,  longitude ],
                  popup=text,
                  icon=folium.Icon(icon='home', color='green'), ).add_to(carte2)

    rayon = 30 # ce n'est bien sur pas la bonne réponse                            car celle-ci doit dépendre de la variable population 
    
    folium.CircleMarker(
                        location=[latitude, longitude],
                        radius=rayon,
                        popup=text,
                        fill=True,
                        fill_color='Red').add_to(carte2)
    
carte2.save('maCarte2.html') # pour sauvegarder la carte
carte2 

In [11]:
# correction

carte2= folium.Map(location=[0, 0],tiles='Stamen Terrain', zoom_start=2)

for i in range(1,21):
    ville = data[i][0]
    population = int (data[i][4]) # int pour avoir un nombre
    text = ville + '  ' + str(population) # str pour avoir du texte à écrire
    
    latitude = data[i][ 1 ]
    longitude = data[i][ 2 ]
    folium.Marker( [latitude,  longitude ],
                  popup=text,
                  icon=folium.Icon(icon='home', color='green'), ).add_to(carte2)

    rayon = population/1000000 # ce n'est bien sur pas la bonne réponse                            car celle-ci doit dépendre de la variable population 
    
    folium.CircleMarker(
                        location=[latitude, longitude],
                        radius=rayon,
                        popup=text,
                        fill=True,
                        fill_color='Red').add_to(carte2)
    
carte2.save('maCarte2.html') # pour sauvegarder la carte
carte2 

<div class="alert alert-success" role="alert">
 <strong> Le travail est terminé !</strong> <br>
</div>